\documentclass[10pt,a4paper]{article}
\usepackage[latin1]{inputenc}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{amssymb}
\usepackage{graphicx}
\title{Manual}

\begin{document}

\maketitle

\section*{Provided files}
\subsection*{summary.xlsx}
A summary excel sheet containing the detected open reading frames (ORFs) together with some additional information. This file contains: 
\begin{itemize}
\item \textbf{ORF ID:} composed of the genome accession, the start, the stop and the strand of the ORF.
\item \textbf{start:} the start position of the ORF
\item \textbf{stop:} the stop position of the ORF
\item \textbf{strand:} the strand of the ORF (+:forward, -:reverse)
\item \textbf{length:} the length of the ORF
\item \textbf{RPKM:} reads per kilobase million measures for each method-condition-replicate triplet
\item \textbf{evidence:} list of 'prediction tool'-'sample' combinations describing the tool by which the ORF was detected and the sample in which the ORF was detected.
\item \textbf{annotated:} if the ORF is already annotated in the provided annotation file, the respective locus tag is shown in this column
\item \textbf{name:} if the ORF is already annotated in the provided annotation file, the respective name (if available) is shown in this column
\item \textbf{ORF type:} ORF type retrieved from reparation (if available)
\item \textbf{start codon:} the start codon for the ORF
\item \textbf{stop codon:} the stop codon for the ORF
\item \textbf{nucleotide sequence:} nucleotide sequence for the ORF
\item \textbf{amino acid sequence:} amino acid sequence for the ORF
\end{itemize}

\subsection*{multi\_qc.html}
The multiQC report collects information from different tools, including fastQC and subread featurecounts.
The general statistics gives an overview over:
\begin{itemize}
\item the number of duplicates
\item the GC content
\item the average read lengths
\item the number of reads (in millions)
\end{itemize}
These statistics are collected after each processing step of our pipeline. 
\begin{itemize}
\item \textbf{raw:} the unprocessed data
\item \textbf{trimmed:} the data after trimming the adapter sequences
\item \textbf{mapped:} the data after mapping with Segemehl
\item \textbf{unique:} the data after removing multi-mapping reads
\item \textbf{norRNA:} the data after filtering the rRNA
\end{itemize}
Further, feature counts are provided for different features from the annotation file. (i.e. how many reads map to each feature)
This includes, all(featurecount), rRNA, norRNA(after filtering), tRNA, ncRNA.

Following is a fastQC report including sequence counts, sequence quality histograms, per sequence quality scores, per base sequence content, per sequence GC content, per base N content, sequence length distribution, sequence duplication levels, over-represented features and adapter content.

\subsection*{heatmap\_SpearmanCorr\_readCounts.pdf}
Spearman correlation coefficients of read counts. The dendrogram indicates which samples read counts are most similar to each other.
\subsection*{genome.fa}
The genome file used for conducting the analysis.
\subsection*{annotation.gff}
The annotation file used for conducting the analysis.
\subsection*{annotation\_rpkm.xlsx}
An excel sheet containing information about the annotated features (taken from the provided annotation). This file contains: 
\begin{itemize}
\item \textbf{Genome:} the genome accession 
\item \textbf{Source:} the source of the gene annotation
\item \textbf{Feature:} the feature of the gene
\item \textbf{Start:} the start position of the annotated gene
\item \textbf{Stop:} the stop position of the annotated gene
\item \textbf{Strand:} the strand of the annotated gene (+:forward, -:reverse)
\item \textbf{Locus tag:} the locus tag of the annotated gene (if available)
\item \textbf{Name:} the name of the annotated gene (if available)
\item \textbf{Length:} the length of the annotated gene
\item \textbf{Product:} the product information for the annotated gene (if available)
\item \textbf{Note:} the notes for the annotated gene (if available)
\item \textbf{RPKM:} reads per kilobase million measures for each method-condition-replicate triplet
\end{itemize}
%TODO: Missing info about raw, min, mil
\subsection*{unfilteredtracks}
A folder containing single nucleotide mapping bigwig files for the data before removal of reads mapping to rRNA. Used for genome browser visualization.
\subsection*{globaltracks}
A folder containing single nucleotide mapping bigwig files for the data after removal of reads mapping to rRNA. Used for genome browser visualization.
\subsection*{threeprimetracks}
A folder containing single nucleotide mapping bigwig files for the data before removal of reads mapping to rRNA. For the region around the three prime end. Used for genome browser visualization.
\subsection*{fiveprimetracks}
A folder containing single nucleotide mapping bigwig files for the data before removal of reads mapping to rRNA. For the region around the five prime end. Used for genome browser visualization.
\subsection*{centeredtracks}
A folder containing single nucleotide mapping bigwig files for the data before removal of reads mapping to rRNA. For the region around the center. Used for genome browser visualization.
\subsection*{tracks}
This folder contains further tracks for genome browser visualization. (colored tracks not yet completed)
\subsubsection{potentialStartCodons.gff}
A genome browser track with possible start codons.
\subsubsection{potentialStopCodons.gff}
A genome browser track with possible stop codons.
\subsubsection{potentialRibosomeBindingSite.gff}
A genome browser track with possible ribosome binding sites.
\subsubsection{potentialAlternativeStartCodons.gff}
A genome browser track with alternative start codons.
\subsubsection{combined.gff}
A genome browser track with the results of the ORF detection tools.
\subsubsection{combined\_annotated.gff}
A genome browser track with the results of the ORF detection tools. Annotated using the existing annotation. (recommended for use)
%\section*{Available files}

\end{document}